src/PrinterCartridgeLevelAlert.php

```diff
@@ -0,0 +1,383 @@
+                'INNER JOIN'   => [
+                    'glpi_cartridgeitems AS i'  => [
+                        'ON'  => [
+                            'c' => 'cartridgeitems_id',
+                            'i'  => 'id'
+                        ]
+                    ],
+                    'glpi_printers_cartridgeinfos AS l'  => [
+                        'ON'  => [
+                            'i' => 'type_tag',
+                            'l'  => 'property'
+                        ]
+                    ],
+                    'glpi_printers as p'  => [
+                        'ON'  => [
+                            'c' => 'printers_id',
+                            'p'  => 'id'
+                        ]
+                    ]
+                ],
+                'LEFT JOIN'    =>  [
+                    'glpi_alerts AS a'  => [
+                        'ON'  => [
+                            'c'  => 'id',
+                            'a'  => 'items_id', [
+                                'AND' => [
+                                    'a.itemtype'    =>      'Cartridge'
+                                ]
+                            ]
+                        ]
+                    ]
+                ],
+                'WHERE'        => $WHERE,
+                'ORDERBY'      => [
+                    'p.name'
+                ]
+            ];
+
+            return $query;
+        } else {
+            return "";
+        }
+    }
+
+    /**
+    * @param $data
+    *
+    * @return string
+    */
+    public static function prepareBodyValues($data)
+    {
+        global $CFG_GLPI;
+
+        $cartridge = new CartridgeItem();
+        $cartridge->getFromDB($data["cartridgeitem"]);
+
+        $printer = new Printer();
+        $printer->getFromDB($data["printer"]);
+
+        $result['cartridge.printer'] = "<a href=\"" . $CFG_GLPI["url_base"] . "/front/printer.form.php?id=" . $printer->fields["id"] . "\">" . $printer->fields["name"];
+
+        if ($_SESSION["glpiis_ids_visible"] == 1 || empty($printer->fields["name"])) {
+            $result['cartridge.printer'] .= " (";
+            $result['cartridge.printer'] .= $printer->fields["id"] . ")";
+        }
+        if (Session::isMultiEntitiesMode()) {
+            $result['cartridge.entity'] = Dropdown::getDropdownName("glpi_entities", $printer->fields["entities_id"]);
+        }
+
+        $result['cartridge.item'] = "<a href=\"" . $CFG_GLPI["url_base"] . "/front/cartridgeitem.form.php?id=" . $cartridge->fields["id"] . "\">" . $cartridge->fields["name"] . " (" . $cartridge->fields["ref"] . ")</a>";
+
+        $result['cartridge.level'] = $data["cartridgelevel"] . (preg_match('#^[0-9]+$#', $data["cartridgelevel"]) ? "%" : "");
+        $result['cartridge.alertDate'] = $data["alertDate"];
+        return $result;
+    }
+
+    /**
+    * @param $data
+    *
+    * @return string
+    */
+    private static function displayBody($data)
```

Comment:
```suggestion
    * @param array $data
    *
    * @return string
    */
    private static function displayBody(array $data): string
```

Reply:
The method name could be confusing since it returns the HTML rather than displaying it.